payroll-processor-client/src/app/data/data-service.ts

```diff
@@ -12,16 +13,17 @@ export class DataService {
   private employees: Employee[] = [];
   private payroll: Payroll[] = [];
 
-  constructor(private http: HttpClient) {
+  private readonly apiUrl: string;
```

Comment:
For any class fields that won't be changed, I mark them `readonly`.

This helps, as classes grow in complexity, to show exactly which state can change. By limiting what state can change we limit that growth in complexity.

Reply:
I like this, currently a standard practice in my C# code but didn't know it was a feature for properties in TypeScript

Reply:
We can also do this with the injected services.

```typescript
private readonly http: HttpClient
```